praw/models/reddit/subreddit.py

```diff
@@ -669,23 +684,86 @@ def _upload_media(
             "mp4": "video/mp4",
             "jpg": "image/jpeg",
             "jpeg": "image/jpeg",
             "gif": "image/gif",
-        }.get(
-            file_extension, "image/jpeg"
-        )  # default to JPEG
+        }
+        if media_path is not None and media_fp is None:
+            if isfile(media_path):
+                file_name = basename(media_path).lower()
+                file_extension = file_name.rpartition(".")[2]
+                mime_type = mime_type.get(
+                    file_extension, "image/jpeg"
+                )  # default to JPEG
+            else:
+                raise TypeError("media_path does not reference a file.")
+        elif media_path is None and media_fp is not None:
+            if isinstance(media_fp, bytes):
+                magic_number = [
+                    int(aByte) for aByte in media_fp[:8]
+                ]  # gets the format indicator
+                file_headers = {
+                    tuple(
+                        [
+                            int(aByte)
+                            for aByte in bytes(
+                                [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]
+                            )
+                        ]
+                    ): "png",
+                    tuple(
+                        [int(aByte) for aByte in bytes([0x6D, 0x6F, 0x6F, 0x76])]
+                    ): "mov",
+                    tuple(
+                        [
+                            int(aByte)
+                            for aByte in bytes(
+                                [0x66, 0x74, 0x79, 0x70, 0x69, 0x73, 0x6F, 0x6D]
+                            )
+                        ]
+                    ): "mp4",
+                    tuple(
+                        [int(aByte) for aByte in bytes([0xFF, 0xD8, 0xFF, 0xE0])]
+                    ): "jpg",
+                    tuple(
+                        [
+                            int(aByte)
+                            for aByte in bytes(
+                                [0xFF, 0xD8, 0xFF, 0xE0, 0x00, 0x10, 0x4A, 0x46]
+                            )
+                        ]
+                    ): "jpeg",
+                    tuple(
+                        [
+                            int(aByte)
+                            for aByte in bytes([0x47, 0x49, 0x46, 0x38, 0x37, 0x61])
+                        ]
+                    ): "gif",
+                }
+                for size in range(4, 10, 2):  # size will equal 4, 6, 8
+                    file_extension = file_headers.get(tuple(magic_number[:size]))
+                    if file_extension is not None:
+                        mime_type = mime_type.get(
+                            file_extension, "image/jpeg"
+                        )  # default to JPEG
+                        file_name = (
+                            mime_type.split("/")[0] + "." + mime_type.split("/")[1]
+                        )
+                        break
```

Comment:
This is the section I'm referring to. Why does this need to be here if fp only contains a file-like object that simply can be read in?

Reply:
That was because the `post` function which was already present requires this `data` parameter to fire:
`upload_response = self._reddit.post(url, data=img_data)`

In this case, `img_data` is being supplied, and it requires the mime_type as a value in its dictionary:
`img_data = {"filepath": file_name, "mimetype": mime_type}`

I wasn't aware/sure of any way to retrieve the `mimetype` from the bytes through an already existing function in any library already included in PRAW, so I just wrote some code to do it manually. If some function already exists to do the following:

```
file_extension = media_fp.get_format() # all of that code could be reduced into this one line with a function of this sort
mime_type = mime_type.get(
                            file_extension, "image/jpeg"
                        )  # default to JPEG
```

I would be more than happy to use that instead. Alternatively there might be a way to not even need the `mimetype` for bytes submissions? I hadn't actually checked since I'm so new to PRAW as a codebase. Let me go check if that's possible...

Reply:
Updating. Presumably it does, but I can't seem to figure it out. Here's the comment under the ```post()``` function:

```"""Return parsed objects returned from a POST request to ``path``.

        :param path: The path to fetch.
        :param data: Dictionary, bytes, or file-like object to send in the body of the
            request (default: ``None``). # This guy is who we want.
        ...
        """
```

Presumably the ```data``` parameter accepts bytes (which is the format I'm trying to submit instead of the traditional format:
```img_data = {"filepath": file_name, "mimetype": mime_type}```) but I can't for the life of me figure it out--though that might be because I'm tired, but I doubt it. I'm not seeing any documentation anywhere on how Reddit handles its API calls and the PRAW code isn't helping. Since the only references online tell me to review PRAW code, I'm effectively stuck in a loop. 

So the gist is, as far as I can tell, if there's a way to submit ```media_fp``` directly to the ```post()``` function so that ```upload_response``` value will be valid going forward, that entire code block you referenced could just be deleted / reworked. Do you know how to accomplish that?